src/texstudio.cpp

```diff
@@ -4413,6 +4422,7 @@ void Texstudio::updateStructure(bool initial, LatexDocument *doc, bool hidden)
 	}
 
 	//structureTreeView->reset();
+	tocTreeView->updateTOC();
```

Comment:
if I remember correctly, this is called with every change of structure.
It updates **all** documents from start to finish (rereading files). This does not scale well ...